lib/charms/mongodb/v0/mongo.py

```diff
@@ -0,0 +1,238 @@
+"""Code for interactions with MongoDB."""
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+import abc
+import logging
+import re
+from dataclasses import dataclass
+from typing import List, Optional, Set
+
+from pymongo import MongoClient
+from pymongo.errors import OperationFailure, PyMongoError
+from tenacity import RetryError, Retrying, stop_after_delay, wait_fixed
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+class NotReadyError(PyMongoError):
+    """Raised when not all replica set members healthy or finished initial sync."""
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3037662a76cc4bf1876d4659c88e77e5"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# path to store mongodb ketFile
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class MongoConfiguration:
```

Comment:
To ensure that this class is not instantiated by itself, what you can do is the following:
```python3
@dataclass 
class AbstractDataclass(ABC): 
    def __new__(cls, *args, **kwargs): 
        if cls == AbstractDataclass or cls.__bases__[0] == AbstractDataclass: 
            raise TypeError("Cannot instantiate abstract class.") 
        return super().__new__(cls)

@dataclass 
class MongoConfiguration(AbstractDataclass):
	...
```

It's not really a must have here as we don't have many developers but it will prevent some errors.

Reply:
:brain: